### ISSUE TITLE
Utilities for object detection

### ISSUE THREAD

**Issue by old-school-kid** [keras-team/keras-cv#96]
Some ready to use utilitilites that could be useful during an object detection application:
- [ ] Label Encoder to transform raw labels into targets for training
- [ ] Anchor Box generation
- [ ] Conversion of anchor box from `[xmin, ymin, xmax, ymax]` to `[x, y, width, height]` and vice versa
- [ ] Anchor optimization for custom data set using [MetaAnchor](https://arxiv.org/abs/1807.00980) (citations 100) or KNN means
- [ ] Visualization of Anchor Box

Not including other utilities like `non max suppression` and `generating bounding box` since they are available under `tf.image`.

**Comment by bhack** [comment]
We could  also double check what useful bbox ops could be copied/refactored from:

https://github.com/tensorflow/models/blob/master/official/vision/beta/ops/box_ops.py

**Comment by old-school-kid** [comment]
> We could also double check what useful bbox ops could be copied/refactored from:

https://github.com/tensorflow/models/blob/master/official/vision/beta/ops/box_ops.py

Thank you @bhack . Skimmed through the code and the first three tasks can be refactored from the code. Moreover `top_k_boxes` and `bbox_matching` from the code could prove to be useful too.

**Comment by bhack** [comment]
Others are in:

https://github.com/tensorflow/models/tree/master/official/vision/utils/object_detection

**Comment by bhack** [comment]
For the generation? 
You could see:
https://github.com/tensorflow/models/blob/master/official/vision/beta/ops/anchor_generator.py
https://github.com/tensorflow/models/blob/master/official/vision/beta/ops/anchor.py

**Comment by innat** [comment]
Weighted box fusion, an alternate of non-max impression. 

- [paper](https://arxiv.org/abs/1910.13302)
- [official code](https://github.com/ZFTurbo/Weighted-Boxes-Fusion)
- [tf code](https://github.com/google/automl/blob/master/efficientdet/tf2/wbf.py)

**Comment by old-school-kid** [comment]
> For the generation? You could see: https://github.com/tensorflow/models/blob/master/official/vision/beta/ops/anchor_generator.py https://github.com/tensorflow/models/blob/master/official/vision/beta/ops/anchor.py

I was talking about optiization of the hyperparameters that doesnt come integrated with tensorflow object detection API. [This repo](https://github.com/joydeepmedhi/Anchor-Boxes-with-KMeans) uses K means whereas MetaAnchor is SOTA afaik.

**Comment by bhack** [comment]
Ah ok.. with enough resources we could review also this.  I suppose with that current PR/maintainer ratio it will be still not a problem.
When this repo will have more popularity probably and PR are going to have lags in the review probably we could try to figure out some priorities.

In that case we could evaluate how much coverage we want to have to all the features related to anchor based models:
[Bridging the Gap Between Anchor-based and Anchor-free Detection via
Adaptive Training Sample Selection](https://openaccess.thecvf.com/content_CVPR_2020/papers/Zhang_Bridging_the_Gap_Between_Anchor-Based_and_Anchor-Free_Detection_via_Adaptive_CVPR_2020_paper.pdf)(CVPR 2020 - citations >300)

**Comment by LukeWood** [comment]
> Some ready to use utilities that could be useful during an object detection application:
> 
> * [ ]  Label Encoder to transform raw labels into targets for training
> * [ ]  Anchor Box generation
> * [x]  Conversion of anchor box from `[xmin, ymin, xmax, ymax]` to `[x, y, width, height]` and vice versa
> * [ ]  Anchor optimization for custom data set using [MetaAnchor](https://arxiv.org/abs/1807.00980) (citations 100) or KNN means
> * [ ]  Visualization of Anchor Box
> 
> Not including other utilities like `non max suppression` and `generating bounding box` since they are available under `tf.image`.

We could also consider adding some sort of a `non max suppression` layer.  The object detection API will definitely take a lot of design.

**Comment by old-school-kid** [comment]
>  The object detection API will definitely take a lot of design.
Yeah, agreed. Should we open a discussion for a that too?

> We could also consider adding some sort of a `non max suppression` layer.
Yeah, @innat provided a nice such example above.

**Comment by LukeWood** [comment]
I’m also considering offering a bounding box extension type:

https://www.tensorflow.org/guide/extension_type

I’m not sure how useful this would be though compared to just using -1 padded tensors or ragged tensors.

**Comment by LukeWood** [comment]
> > The object detection API will definitely take a lot of design.
> 
> Yeah, agreed. Should we open a discussion for this too?
> 
> > We could also consider adding some sort of a `non max suppression` layer.
> 
> Yeah, @innat provided a nice such example above.

Are non-admin users able to open discussions right now?

**Comment by bhack** [comment]
Yes but admin can convert also an issue ti discussion:
https://docs.github.com/en/discussions/managing-discussions-for-your-community/moderating-discussions#converting-an-issue-to-a-discussion

**Comment by innat** [comment]
@LukeWood could you please convert this [issue](https://github.com/keras-team/keras-cv/issues/44) to the discussion too? I think it's better suited there.
